libtrap/src/trap.c

```diff
@@ -3169,50 +3098,59 @@ int trap_ctx_get_data_fmt(trap_ctx_t *ctx, uint8_t ifc_dir, uint32_t ifc_idx, ui
    trap_ctx_priv_t *c = ctx;
 
    if (ctx == NULL) {
-      return TRAP_E_NOT_INITIALIZED;
+      return trap_error(ctx, TRAP_E_NOT_INITIALIZED);
    }
 
    if (ifc_dir == TRAPIFC_INPUT) {
       if (ifc_idx >= c->num_ifc_in) {
-         return TRAP_E_BAD_IFC_INDEX;
+         return trap_error(ctx, TRAP_E_BAD_IFC_INDEX);
       }
 
       inifc = &c->in_ifc_list[ifc_idx];
 
+      pthread_mutex_lock(&inifc->ifc_mtx);
       if (inifc->data_type == TRAP_FMT_RAW) {
-         return TRAP_E_BADPARAMS;
+         pthread_mutex_unlock(&inifc->ifc_mtx);
+         return trap_error(ctx, TRAP_E_BADPARAMS);
       }
 
       if (inifc->client_state == FMT_OK || inifc->client_state == FMT_CHANGED) {
          (*data_type) = inifc->data_type;
          if (inifc->data_type != TRAP_FMT_RAW) {
+            /* TODO: Should probably do copy. */
```

Comment:
Yes, it would be better. But than the returned data will point to a memory which needs to be free'd, so it would need changes in the user code, which may be a problem (maybe create an issue for this?).

Reply:
Issue will be created. The solution would probably require changes in the API.